### ISSUE TITLE
The variable num_slots

### ISSUE THREAD

**Issue by hilder-vitor** [tlepoint/homomorphic-simon#4]
Hi!

I imagine you are no longer working on that implementation, so, this question is not a request for code fixing, I just want to understand better the code to know if I am going to break something when changing it...

The class YASHEKey has a field called _num_slots_.

If I understood well, this value is also used as the maximum possible degree of the plaintexts, right? 
In this case, when we are not using batching, we can only create plaintexts of degree equals to one, since  _num_slots_ is set to 1 at the line [146](https://github.com/tlepoint/homomorphic-simon/blob/master/YASHE/YASHEKey.cpp#L146).

I changed that line to _num_slots = params.n_. So, could you please tell me if my understanding is correct and if you think it will break something ?

I am really thankful for your help!

**Comment by tlepoint** [comment]
Hi,

Sorry no you're gonna break something. If batching is possible, the next few lines of codes change the values of num_slots. The maximum value you can batch is the number of factors of the factorization of the cyclotoic polynomial modulo your plaintext space. If you want it to be maximal, you might have to change the plaintext space value. I defer you to the paper on YASHE (http://eprint.iacr.org/2013/075).

Best,

**Comment by hilder-vitor** [comment]
Hi.

Well, maybe the problem is that I don't understand how the BitVector class models a polynomial. I thought it was like that:

> each position _k_ of a BitVector corresponded to a coefficient _a_k_. 

In that case, the degree of the polynomial would be _BitVector::l()_.

But the decryption algorithm (_YASHEKey::decrypt_) initializes a BitVector passing the _num_slots_ to its constructor. Therefore, if it models a polynomial as I said above, YASHEKey::decrypt creates a polynomial of degree zero if batching is not being used, because  num_slots is one in this case.

Thank you very much for answering!

**Comment by tlepoint** [comment]
Hi,

Well I linked to the initial YASHE paper, not mine ;-) 

Anyway, the BitVector class does not represent a polynomial, but the messages. You want to encrypt a message, it is either a single element (num_slots=1) or a vector of elements (num_slots>0). Whatever the case, it holds that num_slots <= deg(cyclotomic polynomial). Then, the message is encrypted either in the constant coefficient of the ciphertext polynomial (num_slots=1) or in the constant coefficients of a CRT base of the cyclotomic polynomial added to the ciphertext polynomial. Decrypting is similar.

In other words, the size of BitVector() is only the plaintext size, not the ciphertext size. Please look at the following paper for more detail: [https://eprint.iacr.org/2012/099](https://eprint.iacr.org/2012/099).

**Comment by hilder-vitor** [comment]
I read again the three papers. The YASHE's plaintext messages are elements of R/tR, which means polynomials with degree less than deg(cyclotomic polynomial) and coefficients in Z_t.

Thus, in my understanding, if I were using batching with `num_slots` slots, I should have `num_slots` polynomials of degree `y` such that `y * num_slots` equals deg(cyclotomic polynomial). Therefore, I would need  deg(cyclotomic polynomial) integers in Z_t, so, a BitVector with length equals to deg(cyclotomic polynomial).

And if I don't want to use the batching technique, the BitVector should model a single element of R/tR, but it still should have length equals to deg(cyclotomic polynomial).

This is the reason I don't understand why the BitVector is initialized as `BitVector m(num_slots);` in  [YasheKey::decrypt](https://github.com/tlepoint/homomorphic-simon/blob/master/YASHE/YASHEKey.cpp#L292).    :(

Well, at least now it is clear why the assignment  `num_slots = params.n` doesn't make sense.

**Comment by tlepoint** [comment]
Ok I understand what puzzles you.

The implementation here considers **integer** plaintext, i.e. either Z_t without batching, or Z_t^d with batching for d | phi(n) according to the factorization of the cyclotomic polynomial.
If you want to implement **polynomial** plaintexts, you have to modify the encrypt/decrypt procedures accordingly.

Best,

**Comment by hilder-vitor** [comment]
Ah, ok, I got it.   :smiley: 

Again, thank you very much for helping!
